### PR TITLE
Fix "!=" and "not ilike" failing to match null records

### PR DESCRIPTION
**Pull request by xmo-odoo**

### CODE REVIEW

File: odoo/osv/expression.py [hunk 1]
```diff
@@ -1265,14 +1265,16 @@ def __leaf_to_sql(self, eleaf):
...
-                    query = '(%s OR %s."%s" IS NULL)' % (query, table_alias, left)
-                params = ['%%%s%%' % native_str]
+                params = ['%%%s%%' % pycompat.to_text(right)]
             else:
                 field = model._fields[left]
                 params = [field.convert_to_column(right, model, validate=False)]
 
+            # what should the result be if left is null?
+            null_as = bool(right) == (operator in NEGATIVE_TERM_OPERATORS)
+            query = 'coalesce(%s, %%s)' % query
```
Comment (Yajo) [inline odoo/osv/expression.py]: I'm just guessing, so I might be saying something silly, but wouldn't it be more performant and maintainable something like this pseudocode?:

```py
if operator in NEGATIVE_TERM_OPERATORS and right is not False:
    query = "(%s OR %s IS NULL)" % (query, left)
```

... this way you'd avoid the need to provide a proper coalesced value depending on the column type (i.e. an int should be coalesced to 0, a str to "", a float to 0.0, a bool to False, etc.).

Also I imagine it's faster for postgres the `IS NOT NULL` than the `COALESCE(*,*) == whatever`.

Comment (xmo-odoo) [reply odoo/osv/expression.py]: Not at all silly, this was and is very much an option. The original was an attempt to make the code more maintainable by having less branches / conditions (and even removing existing ones).

Comment (odony) [reply odoo/osv/expression.py]: @xmo-odoo Did you run any comparison between `COALESCE` and `OR IS NULL` explain plans, to get some supporting evidence?

Comment (xmo-odoo) [reply odoo/osv/expression.py]: @odony didn't reach that point, the coalesce version broke some bits so I went with the conditional IS NULL which seem to work.

**Comment by pedrobaeza** [comment]
@xmo-odoo is this OK to be merged? I don't want this to be forgotten before v13 freeze. :pray:

**Comment by odony** [comment]
@robodoo r+ 
@xmo-odoo I particularly love the PR description :+1:

**Pull request closed without merging**